drivers/bus/acpi/eval.c

```diff
@@ -79,177 +91,384 @@ ConvertPackageArguments(ACPI_METHOD_ARGUMENT *Argument,
+    ULONG ExtraParamLength, OutputBufSize;
+    PACPI_EVAL_OUTPUT_BUFFER OutputBuf;
+    NTSTATUS Status;
+    ULONG Count = 1;
+
+    /* If we didn't get anything back then we're done */
+    if (!ReturnBuffer->Pointer || ReturnBuffer->Length == 0 ||
+        IoStack->Parameters.DeviceIoControl.OutputBufferLength == 0)
+    {
+        return STATUS_SUCCESS;
+    }
+
+    Obj = ReturnBuffer->Pointer;
+    switch (Obj->Type)
+    {
+        case ACPI_TYPE_INTEGER:
+            ExtraParamLength = sizeof(ULONG);
+            break;
+
+        case ACPI_TYPE_STRING:
+            ExtraParamLength = Obj->String.Length + sizeof(UCHAR);
+            break;
+
+        case ACPI_TYPE_BUFFER:
+            ExtraParamLength = Obj->Buffer.Length;
+            break;
+
+        case ACPI_TYPE_PACKAGE:
+            Status = EvalGetPackageSize(Obj, &Count, &ExtraParamLength);
+            if (!NT_SUCCESS(Status))
+                return Status;
+            break;
+
+        default:
+           ASSERT(FALSE);
+           return STATUS_UNSUCCESSFUL;
+    }
+
+    if (!AcpiVerifyOutBuffer(IoStack, sizeof(ACPI_EVAL_OUTPUT_BUFFER)))
+    {
+        DPRINT1("Buffer too small\n");
+
+        return STATUS_BUFFER_TOO_SMALL;
+    }
+
+    OutputBufSize = FIELD_OFFSET(ACPI_EVAL_OUTPUT_BUFFER, Argument) +
+                    ACPI_METHOD_ARGUMENT_LENGTH(ExtraParamLength);
+
+    OutputBuf = Irp->AssociatedIrp.SystemBuffer;
+    OutputBuf->Signature = ACPI_EVAL_OUTPUT_BUFFER_SIGNATURE;
+    OutputBuf->Length = OutputBufSize;
+    OutputBuf->Count = Count;
+
+    if (!AcpiVerifyOutBuffer(IoStack, OutputBufSize))
+    {
+        DPRINT1("Buffer too small %u/%u\n",
+                IoStack->Parameters.DeviceIoControl.OutputBufferLength,
+                OutputBufSize);
+
+        Irp->IoStatus.Information = OutputBufSize;
+        return STATUS_BUFFER_OVERFLOW;
+    }
+
+    switch (Obj->Type)
+    {
+        case ACPI_TYPE_INTEGER:
+            ACPI_METHOD_SET_ARGUMENT_INTEGER(OutputBuf->Argument, Obj->Integer.Value);
+            break;
+
+        case ACPI_TYPE_STRING:
+            ACPI_METHOD_SET_ARGUMENT_STRING(OutputBuf->Argument, Obj->String.Pointer);
+            break;
+
+        case ACPI_TYPE_BUFFER:
+            ACPI_METHOD_SET_ARGUMENT_BUFFER(OutputBuf->Argument,
+                                            Obj->Buffer.Pointer,
+                                            Obj->Buffer.Length);
+            break;
+
+        case ACPI_TYPE_PACKAGE:
+            Status = EvalConvertPackageArguments(OutputBuf->Argument, Obj);
+            if (!NT_SUCCESS(Status))
+                return Status;
+            break;
+
+        default:
+            ASSERT(FALSE);
+            return STATUS_UNSUCCESSFUL;
+    }
+
+    Irp->IoStatus.Information = OutputBufSize;
+    return STATUS_SUCCESS;
+}
 
 NTSTATUS
 NTAPI
-Bus_PDO_EvalMethod(PPDO_DEVICE_DATA DeviceData,
-                   PIRP Irp)
+Bus_PDO_EvalMethod(
+    PPDO_DEVICE_DATA DeviceData,
+    PIRP Irp)
```

Comment:
SAL annotations

Reply:
This function is an external symbol and the driver is long known to not contain any SAL annotations.